[MTKA-1479] refactor post slug generation

mindctrl opened the pull request:
## Description

- [x] Continue generating fallback slug on posts with no title field
- [x] Generate post slug from title field
- [x] Tests

<!--
Include a summary of the change and some contextual information.
-->

<!--
Provide a link to the JIRA ticket (if any) for issue tracking purposes
-->

https://wpengine.atlassian.net/browse/MTKA-1479

## Testing

<!--
Describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Also list any relevant details for your test configuration such as how to test the changes locally or in staging.
-->

## Screenshots

<!--
If this is a visual change include relevant screenshots about the behavior of the application before and after this change.
-->

## Documentation Changes

<!--
List corresponding changes to the documentation and updated wiki pages.
-->

## Dependant PRs

<!--
List any dependent PR's that are awaiting review. And in order to have these dependencies listed as part of the checks section, use the following syntax:

Depends on #1234
-->

mindctrl committed:
Merge branch 'feature/post-titles' into fix/post-slugs

mindctrl committed:
test: check post ID value instead of slug when confirming post visibility in API

mindctrl committed:
test: check post title matches mock title value

mindctrl committed:
refactor: adjust post name generation in `set_post_attributes()`

mindctrl committed:
refactor: set default post_name and post_title values on auto-draft posts

mindctrl committed:
chore: remove `set_post_attributes` callback from `wp_insert_post` hook.

Prevents the callback from running during post updates that save post title.

mindctrl committed:
chore: revert nonce check addition.

mindctrl committed:
chore: add auto-draft entry to test post data

mindctrl committed:
test: confirm default post title value for auto-draft posts

mindctrl committed:
test: confirm post name values for published and auto-draft posts

nickcernis commented on includes/publisher/class-publisher-form-editing-experience.php:
```diff
@@ -220,7 +220,9 @@ public function enqueue_assets( string $hook ): void {
 						$value = get_post_meta( $post->ID, $field['slug'], true );
 						if ( ! empty( $field['isTitle'] ) && $value !== $post->post_title ) {
 							$post->post_title = $value;
+							remove_action( 'wp_insert_post', [ $this, 'set_post_attributes' ] );
```
Might it be worth explaining inline why we're removing and re-adding the action here, either as a comment or by extracting this repeated logic as a small function?

mindctrl replied:
Extracted in https://github.com/wpengine/atlas-content-modeler/pull/502/commits/92e4897184116a9727aad9a92bdd3894edc9438b.

nickcernis reviewed (commented):
@mindctrl Right now I see the slug change when I change the title field value.

Is that expected? (It's not how post/page title editing works — the slug is constant after publishing unless you explicitly edit it; I can edit the title of a post/page but it's still accessible at the original slug.)

Not a blocker but it might make slugs unstable if queries expect them not to change unexpectedly, so it seemed worth mentioning.

mindctrl commented:
@nickcernis good catch. Not intended. I think I broke it refactoring some bits before pushing up. Will take another look.

mindctrl committed:
refactor: only set fallback post titles and slugs on auto-draft posts.

mindctrl committed:
test: confirm post title value for auto-draft posts

mindctrl committed:
refactor: extract duplicated logic to private method

mindctrl committed:
chore: update post slug when copying title field values

mindctrl committed:
test: confirm post name values for published and auto-draft posts

mindctrl committed:
test: use model with fields when testing for field values

mindctrl committed:
chore: improve separation of title and slug tests

mindctrl committed:
test: re-fetch title value from db to ensure we're testing the right value.

mindctrl committed:
test: confirm updating title value does not update post_name.

We only generate a post_name from the title when the post_name
is either empty or is the same as the post ID number.

nickcernis reviewed (approved):
Lookin' good! I manually confirmed that:

- [x] Models with a title field copy title to post slug on save.
- [x] Models without a title field use the post ID for the slug.
- [x] Editing an existing post to change the title field content does not change the slug.
